ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -117,76 +132,42 @@ delayedProgressReporting before after (Just lspEnv) optProgressStyle = do
     progressState <- newVar NotStarted
     let progressUpdate event = updateStateVar $ Event event
         progressStop   =  updateStateVar StopProgress
-        updateStateVar = modifyVar_ progressState . updateState (lspShakeProgress inProgressState)
-
+        updateStateVar tran = do
+            start <- lspShakeProgressNew inProgressState
+            modifyVar_ progressState $ updateState start tran
         inProgress = updateStateForFile inProgressState
     return ProgressReporting{..}
     where
-        lspShakeProgress InProgressState{..} = do
+        lspShakeProgressNew InProgressState{..} = do
             -- first sleep a bit, so we only show progress messages if it's going to take
             -- a "noticable amount of time" (we often expect a thread kill to arrive before the sleep finishes)
             liftIO $ sleep before
-            u <- ProgressToken . InR . T.pack . show . hashUnique <$> liftIO newUnique
-
-            b <- liftIO newBarrier
-            void $ LSP.runLspT lspEnv $ LSP.sendRequest SMethod_WindowWorkDoneProgressCreate
-                LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
-            liftIO $ async $ do
-                ready <- waitBarrier b
-                LSP.runLspT lspEnv $ for_ ready $ const $ bracket_ (start u) (stop u) (loop u 0)
+            cancelProgress <- newEmptyMVar
+            LSP.runLspT lspEnv $ do
+                u <- ProgressToken . InR . T.pack . show . hashUnique <$> liftIO newUnique
+                b <- liftIO newBarrier
+                LSP.sendRequest SMethod_WindowWorkDoneProgressCreate
+                    LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
+                async $ do
+                    ready <- liftIO $ waitBarrier b
+                    withProgress "Processing" (Just u) Cancellable $ \update -> loopUntil cancelProgress update 0
```

Comment:
This isn't quite right, you should skip sending the `WorkDoneProgressCreate` message and pass `Nothing` here. Then it creates the token for you automatically. Passing `Just token` is for the case where you can get a progress token in the request, which isn't what's happening here.

Reply:
Then you can skip all the other stuff and just go straight to the `withProgress` call.

Also, while we're here, can we change the title to something other than the incredibly generic "Processing"? I've always found that quite annoying. We might need to pass in an informative string from the callers, though...

Reply:
done